### PR TITLE
Rigid Table column definitions 

### LINKED ISSUE #652: Remove the empty initial Chunk and make Table layout immutable
This will also fix #649 

The Table's initial empty Chunk is an architectural confusion. Operators producing no output will have it in their output table and subsequent operators will fail because they assume the Chunk layout to be the same as the Table's - which it isn't. 
**Secondly** and closely related: The fact that you can add column definitions to a Table is another potential source of mistakes, because again Chunk and Table layout may start to differ. I want the Table's layout to be passed in in the constructor and then remain immutable.
**Thirdly** you can add Columns to Chunks after they have been added to a table. Again, you easily (and probably accidentally) reach an invalid Table-Chunk state.

**Table and Chunk layouts shouldn't be allowed to diverge.** 

Does this sound reasonable @mrks @Bensk1?

I propose the following:

```c++
const auto layout = HorizontalTableLayout{}; // Name to be discussed. Maybe just HorizontalLayout since it also goes into the Chunk?
layout.add_column_definition(DataType::Int, "a");
layout.add_column_definition(DataType::Float, "b");

auto table = std::make_shared<Table>(layout);

std::vector<std::shared_ptr<BaseColumn>> columns;
/** ...fill columns with data */

table->create_new_chunk(columns, [allocator, access_counter]);  // Calls Chunk::Chunk(_layout, /**mvcc*/ columns);

// Remove Table::add_column, Table::add_column_defintion, Chunk::add_column
```

### PR DESCRIPTION
**Pull request by mrzzzrm**
Fixes #652, #649

Table's now have a static layout (no more `add_column`/`add_column_definition`) and a static `TableType`. New Chunks get added via `Table::append_chunk` which takes the columns and makes sure a Chunk with the appropriate MVCC setting is created and appended.

Chunk column layout becomes static as well, not allowing the appending of columns after the Chunk has been created. 

Along with the changes for #652, this includes some fixes/refactoring that made it in accidentally or became necessary because a bug surfaced.

 - JoinHash, JoinNestedLoop and JoinSortMerge couldn't handle an empty reference table when performing outer joins. As soon as a table was a reference table, they assumed it had >= 1 chunks and determined the referencED table using this first Chunk. This was fixed by having output (which can just be NULL values) produced from an empty input table to reference a DummyTable. CC @ArneMayer 
 - Whether intended or not, the IsNullTableScan relied on NULL_ROW_ID having a ChunkID of 0, with the code handling the first Chunk also handling NULL_ROW_IDs implicitly and without any commenting mention of this behaviour. Since a table can have 0 chunks now, this needed to be fixed and actually got rid of some null value handling in the iterators. CC @mjendruk 
- The Aggregate operator couldn't handle input tables without chunks, since it would create the AggregateContexts when processing the first chunk. Fixed.
 - MvccColumns and ChunkAccessCounter moved to their own files...

### CODE REVIEW

**Comment by mrks** [comment]
Are there any files that we can safely skip in the review? Or in other words: Which of those 3000 lines do we have to look at?

**Comment by mrzzzrm** [comment]
Bloody good question. 

What just needs glancing over
 - replacing `add_column_definition/add_column` with `TableColumnDefinitions`
 - moving the actual creation of the table further down in the code
 -  replacing `Table::emplace_chunk` with `Table::append_chunk`
 - replacing `std::make_shared<Chunk>` + `chunk->add_column` with `ChunkColumns` + `push_back`
...this is most of the changes.

What should be looked at more thoroughly are the two points in my opening post where I tagged Arne and Max, as well as the changes in the `Table` and `Chunk` implementations. If you agree with those, then all other changes are likely okay with you as well.

File: src/benchmark/table_generator.cpp [hunk 1]
```diff
@@ -20,19 +20,19 @@ namespace opossum {
...
    */
+  TableColumnDefinitions column_definitions;
   for (size_t i = 0; i < _num_columns; i++) {
     auto column_name = std::string(1, static_cast<char>(static_cast<int>('a') + i));
-    table->add_column_definition(column_name, DataType::Int);
+    column_definitions.emplace_back(column_name, DataType::Int);
     value_vectors.emplace_back(tbb::concurrent_vector<int>(vector_size));
   }
-  auto chunk = std::make_shared<Chunk>();
+  std::shared_ptr<Table> table = std::make_shared<Table>(column_definitions, TableType::Data, chunk_size);
```
Comment (mjendruk) [inline src/benchmark/table_generator.cpp]: `auto table = ...`

File: src/benchmarklib/benchmark_utilities/abstract_benchmark_table_generator.hpp [hunk 2]
```diff
@@ -51,18 +51,19 @@ class AbstractBenchmarkTableGenerator {
...
   template <typename T>
-  void add_column(std::shared_ptr<opossum::Table> table, std::string name,
+  void add_column(std::vector<opossum::ChunkColumns>& columns_by_chunk,
+                  opossum::TableColumnDefinitions& column_definitions, std::string name,
                   std::shared_ptr<std::vector<size_t>> cardinalities,
                   const std::function<std::vector<T>(std::vector<size_t>)>& generator_function) {
     /**
      * We have to add Chunks when we add the first column.
      * This has to be done after the first column was created and added,
      * because empty Chunks would be pruned right away.
```
Comment (mjendruk) [inline src/benchmarklib/benchmark_utilities/abstract_benchmark_table_generator.hpp]: Does this comment still apply?

File: src/lib/CMakeLists.txt [hunk 3]
```diff
@@ -379,7 +385,7 @@ if (NOT DISABLE_JIT_SUPPORT)
         ${SOURCES}
         ${JIT_EMBEDDED_SOURCES}
         ${LLVM_BUNDLE_FILE}
-    )
+        storage/mvcc_columns.cpp storage/mvcc_columns.hpp)
```
Comment (mjendruk) [inline src/lib/CMakeLists.txt]: this seems to be a duplicate of lines 267–268

File: src/lib/operators/join_hash.cpp [hunk 4]
```diff
@@ -694,64 +671,71 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+            }
           }
+          output_columns.push_back(std::make_shared<ReferenceColumn>(ref_col->referenced_table(),
+                                                                     ref_col->referenced_column_id(), new_pos_list));
+        } else {
+          // If there are no Chunks in the input_table, we can't deduce the Table that input_table is referencING to
+          // pos_list will contain only NULL_ROW_IDs anyway, so it doesn't matter which Table the ReferenceColumn that
+          // we output is referencing. HACK, but works fine: we create a dummy table and let the ReferenceColumn ref
+          // it.
+          const auto dummy_table = std::make_shared<Table>(input_table->column_definitions(), TableType::Data);
```
Comment (mjendruk) [inline src/lib/operators/join_hash.cpp]: Is this done anywhere else?

File: src/lib/operators/join_nested_loop.cpp [hunk 5]
```diff
@@ -205,54 +207,60 @@ void JoinNestedLoop::_perform_join() {
...
+            std::static_pointer_cast<const ReferenceColumn>(input_table->get_chunk(ChunkID{0})->get_column(column_id));
+
+        column = std::make_shared<ReferenceColumn>(reference_column->referenced_table(),
+                                                   reference_column->referenced_column_id(), new_pos_list);
+      } else {
+        // If there are no Chunks in the input_table, we can't deduce the Table that input_table is referencING to
+        // pos_list will contain only NULL_ROW_IDs anyway, so it doesn't matter which Table the ReferenceColumn that
+        // we output is referencing. HACK, but works fine: we create a dummy table and let the ReferenceColumn ref
+        // it.
+        const auto dummy_table = std::make_shared<Table>(input_table->column_definitions(), TableType::Data);
```
Comment (mjendruk) [inline src/lib/operators/join_nested_loop.cpp]: Okay it is. How about adding a static method to `ReferenceColumn`: `create_with_dummy_table`?

File: src/lib/operators/table_scan.cpp [hunk 6]
```diff
@@ -94,11 +94,12 @@ std::shared_ptr<const Table> TableScan::_on_execute() {
...
       // The actual scan happens in the sub classes of BaseTableScanImpl
       const auto matches_out = std::make_shared<PosList>(_impl->scan_chunk(chunk_id));
+      if (matches_out->empty()) return;
 
-      // The output chunk is allocated on the same NUMA node as the input chunk. Also, the AccessCounter is
+      // The output chunk is allocated on the same NUMA node as the input chunk. Also, the ChunkAccessCounter is
       // reused to track accesses of the output chunk. Accesses of derived chunks are counted towards the
       // original chunk.
-      auto chunk_out = std::make_shared<Chunk>(chunk_guard->get_allocator(), chunk_guard->access_counter());
+      ChunkColumns out_columns;
```
Comment (mjendruk) [inline src/lib/operators/table_scan.cpp]: How do you ensure that what is written above is still true? What did this guard do anyway?

File: src/lib/operators/table_scan/is_null_table_scan_impl.cpp [hunk 7]
```diff
@@ -17,27 +18,44 @@ namespace opossum {
...
+}
+
+void IsNullTableScanImpl::handle_column(const ReferenceColumn& left_column,
+                                        std::shared_ptr<ColumnVisitableContext> base_context) {
+  auto context = std::static_pointer_cast<Context>(base_context);
+  BaseSingleColumnTableScanImpl::handle_column(left_column, base_context);
+
+  const auto pos_list = *left_column.pos_list();
+
+  // Additionally to the null values in the referencED column, we need to find null values in the referencING column
```
Comment (mjendruk) [inline src/lib/operators/table_scan/is_null_table_scan_impl.cpp]: 👍

File: src/lib/storage/chunk.hpp [hunk 8]
```diff
@@ -24,7 +28,7 @@ namespace opossum {
 class BaseIndex;
 class BaseColumn;
 
-enum class ChunkUseAccessCounter { Yes, No };
+using ChunkColumns = pmr_concurrent_vector<std::shared_ptr<BaseColumn>>;
```
Comment (mjendruk) [inline src/lib/storage/chunk.hpp]: Why does it have to be a concurrent vector if chunks are immutable anyway?

File: src/lib/storage/chunk.cpp [hunk 9]
```diff
@@ -49,7 +44,7 @@ void Chunk::replace_column(size_t column_id, std::shared_ptr<BaseColumn> column)
 
 void Chunk::append(const std::vector<AllTypeVariant>& values) {
   // Do this first to ensure that the first thing to exist in a row are the MVCC columns.
-  if (has_mvcc_columns()) grow_mvcc_column_size_by(1u, Chunk::MAX_COMMIT_ID);
```
Comment (mjendruk) [inline src/lib/storage/chunk.cpp]: should we assert `is_mutable` here?

File: src/lib/storage/table.cpp [hunk 10]
```diff
@@ -154,55 +134,36 @@ const ProxyChunk Table::get_chunk_with_access_counting(ChunkID chunk_id) const {
...
+                         const std::shared_ptr<ChunkAccessCounter>& access_counter) {
+  const auto chunk_size = columns.empty() ? 0u : columns[0]->size();
 
-  // We assume if one column is a reference column, all are.
-  const auto column = _chunks[0]->get_column(ColumnID{0});
-  const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(column);
-
-  if (ref_column != nullptr) {
-// In debug mode we're pedantic and check whether all columns in all chunks are Reference Columns
 #if IS_DEBUG
```
Comment (mjendruk) [inline src/lib/storage/table.cpp]: I’d prefer 
```c++
DebugAssert([&]() {
  return true;
}(), "comment");
```

If you keep this style, please replace `DebugAssert` with `Assert` below.

File: src/lib/storage/table.hpp [hunk 11]
```diff
@@ -19,36 +20,59 @@ namespace opossum {
 
 class TableStatistics;
 
-// A table is partitioned horizontally into a number of chunks
+/**
+ * A Table is partitioned horizontally into a number of chunks.
+ *
+ *
```
Comment (mjendruk) [inline src/lib/storage/table.hpp]: A few too many comment lines for my taste.

File: src/lib/storage/table_column_definition.hpp [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  bool operator==(const TableColumnDefinition& rhs) const;
+
+  std::string name;
+  DataType data_type{DataType::Int};
+  bool nullable{false};
+};
+
+using TableColumnDefinitions = std::vector<TableColumnDefinition>;
+
+TableColumnDefinitions concatenated(const TableColumnDefinitions& lhs, const TableColumnDefinitions& rhs);
```
Comment (mjendruk) [inline src/lib/storage/table_column_definition.hpp]: I like this. Could we add another version for joins where one definition is merged?

File: src/lib/storage/table.hpp [hunk 13]
```diff
@@ -19,36 +20,59 @@ namespace opossum {
...
-  // default is the maximum allowed chunk size. A table holds always at least one chunk
-  explicit Table(const uint32_t max_chunk_size = Chunk::MAX_SIZE);
+  const TableColumnDefinitions& column_definitions() const;
+
+  size_t column_count() const;
+
+  const std::string& column_name(const ColumnID column_id) const;
+  std::vector<std::string> column_names() const;
+
+  DataType column_data_type(const ColumnID column_id) const;
```
Comment (mjendruk) [inline src/lib/storage/table.hpp]: Thanks for renaming this. This was also on my list.

File: src/lib/types.hpp [hunk 14]
```diff
@@ -84,13 +86,23 @@ using pmr_ring_buffer = boost::circular_buffer<T, PolymorphicAllocator<T>>;
...
 
 struct RowID {
-  ChunkID chunk_id{0};
+  ChunkID chunk_id{INVALID_CHUNK_ID};
   ChunkOffset chunk_offset{INVALID_CHUNK_OFFSET};
 
+  RowID() = default;
+
+  RowID(const ChunkID chunk_id, const ChunkOffset chunk_offset) : chunk_id(chunk_id), chunk_offset(chunk_offset) {
+    DebugAssert((chunk_offset == INVALID_CHUNK_OFFSET) == (chunk_id == INVALID_CHUNK_ID),
```
Comment (mjendruk) [inline src/lib/types.hpp]: Okay this took me a while. I’d prefer `DebugAssert(*this != NULL_ROW_ID, "");`

File: src/benchmarklib/benchmark_utilities/abstract_benchmark_table_generator.hpp (hunk 2 above)
Comment (mrzzzrm) [reply src/benchmarklib/benchmark_utilities/abstract_benchmark_table_generator.hpp]: well spotted, it doesn't

File: src/lib/CMakeLists.txt (hunk 3 above)
Comment (mrzzzrm) [reply src/lib/CMakeLists.txt]: here is CLion trying to be smarter than I am

File: src/lib/operators/join_nested_loop.cpp (hunk 5 above)
Comment (mrzzzrm) [reply src/lib/operators/join_nested_loop.cpp]: makes it just a tiny bit cleaner. but I went for it

File: src/lib/operators/table_scan.cpp (hunk 6 above)
Comment (mrzzzrm) [reply src/lib/operators/table_scan.cpp]: The "The output chunk is allocated on the same NUMA node as the input chunk" bit is a lie as far as I can tell, look at the place where the pos list is allocated. Removed that part of the comment. I will open an issue requesting the retirement of the current NUMA-awareness, but that's not part of this PR

File: src/lib/storage/chunk.hpp (hunk 8 above)
Comment (mrzzzrm) [reply src/lib/storage/chunk.hpp]: Yeah, I guess than can become a std::vector now.

File: src/lib/storage/chunk.cpp (hunk 9 above)
Comment (mrzzzrm) [reply src/lib/storage/chunk.cpp]: good point

File: src/lib/storage/chunk.hpp (hunk 8 above)
Comment (mrzzzrm) [reply src/lib/storage/chunk.hpp]: Sadly, Chunk::migrate won't work with a normal std::vector

File: src/lib/storage/table.cpp (hunk 10 above)
Comment (mrzzzrm) [reply src/lib/storage/table.cpp]: The current style makes it easier to debug, with a lambda you just know THAT it failed, not WHERE and in what state.

File: src/lib/storage/table.hpp (hunk 11 above)
Comment (mrzzzrm) [reply src/lib/storage/table.hpp]: you think so? tsss....

File: src/lib/types.hpp (hunk 14 above)
Comment (mrzzzrm) [reply src/lib/types.hpp]: Clever

Comment (mrzzzrm) [reply src/lib/types.hpp]: but not doing what I intended. What this Asserts is that "If you pass in one of the arguments as INVALID/NULL, the other has to be INVALID/NULL as well. This makes sure there is just one value representing an invalid row id."

File: src/lib/storage/chunk.hpp (hunk 8 above)
Comment (mjendruk) [reply src/lib/storage/chunk.hpp]: Why shouldn’t it? You just have to change the type there too!

File: src/lib/types.hpp (hunk 14 above)
Comment (mjendruk) [reply src/lib/types.hpp]: Should it ever be possible to use `INVALID_CHUNK_OFFSET` and `INVALID_CHUNK_OFFSET` separately?

File: src/lib/storage/chunk.hpp (hunk 8 above)
Comment (mrzzzrm) [reply src/lib/storage/chunk.hpp]: oh yeah stupid me.

Comment (mrzzzrm) [reply src/lib/storage/chunk.hpp]: wanted to write a sarcastic comment but this suggestion **actually works** if you also remove the PolymorphicAllocator from the ChunkColumns as well i Chunk::migrate(). But I'm not doing that in this PR, even though I'm in favour of this in general.

File: src/lib/types.hpp (hunk 14 above)
Comment (mrzzzrm) [reply src/lib/types.hpp]: Don't know what you are getting at. Maybe? Why forbid it?

File: src/lib/storage/chunk.hpp (hunk 8 above)
Comment (mjendruk) [reply src/lib/storage/chunk.hpp]: Dude, just use `pmr_vector`!

Comment (mrzzzrm) [reply src/lib/storage/chunk.hpp]: I don't even want to get into this allocator nonsense....

File: src/lib/storage/table_column_definition.hpp (hunk 12 above)
Comment (mrzzzrm) [reply src/lib/storage/table_column_definition.hpp]: What do you mean by "merged"? I'm using this function for joins, am I not?

**Comment by mrzzzrm** [comment]
Alright, there is your pmr vector

File: src/lib/storage/table_column_definition.hpp (hunk 12 above)
Comment (mjendruk) [reply src/lib/storage/table_column_definition.hpp]: Nevermind

**Comment by mrzzzrm** [comment]
Now CI has to run again. But I'm certain I want this in @mrks. Will look into Seminar PR code pronto and offer my help with resolving issues/conflicts after merge, promised.

File: src/benchmark/operators/union_positions_benchmark.cpp [hunk 15]
```diff
@@ -57,10 +57,12 @@ class UnionPositionsBenchmarkFixture : public benchmark::Fixture {
      * Create the referenced table, that doesn't actually contain any data - but UnionPositions won't care, it just
      * operates on RowIDs
      */
-    _referenced_table = std::make_shared<Table>();
+    TableColumnDefinitions column_defintions;
```
Comment (mrks) [inline src/benchmark/operators/union_positions_benchmark.cpp]: definitions

**Comment by mrks** [comment]
ok

**Comment by mrzzzrm** [comment]
@mrks quick, before the master conflicts again!

**Comment by mrks** [comment]
I didn't know that you were waiting for me. ;)

**Pull request merged**